yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,590 @@
+
+        # These don't have setters as these would invalidate the tree which is
+        # a feature we don't have
+        self._data_version = data_version
+        self._n_ref = n_ref
+        self._num_particles = input_pos.shape[0]
+        self._max_depth = max_depth
+
+        # setting the properties which determines how children divide
+        self.over_refine_factor = over_refine_factor
+        self.density_factor = density_factor
+
+        # set up the initial idx of the particles, this keeps track of which
+        # particle from the input is in which node
+        self._idx = np.arange(0, input_pos.shape[0], dtype=np.int64)
+
+        # set up the bounds and root node
+        self.setup_bounds(input_pos, left_edge, right_edge)
+        self.setup_root(input_pos)
+
+        # reserve some space for the octs to be stored
+        reserve(&self.nodes, 1000000)
+
+        # now build the tree
+        self.build_tree(&input_pos[0, 0])
+
+    cdef setup_bounds(self, double[:, ::1] &input_pos, left_edge=None,
+                      right_edge=None):
+        if left_edge is not None:
+            for i in range(3):
+                self._left_edge[i] = left_edge[i]
+        else:
+            for i in range(3):
+                self._left_edge[i] = np.amin(input_pos[:,i])
+
+        if right_edge is not None:
+            for i in range(3):
+                self._right_edge[i] = right_edge[i]
+        else:
+            for i in range(3):
+                self._right_edge[i] = np.amax(input_pos[:,i])
+
+    cdef setup_root(self, double[:, ::1] &input_pos):
+        cdef Node root
+        root.left_edge = self._left_edge
+        root.right_edge = self._right_edge
+        root.parent = 0
+
+        root.start = 0
+        root.end = input_pos.shape[0]*3
+
+        root.children = 0
+        root.leaf = 1
+        root.depth = 0
+        root.leaf_id = 0
+        root.node_id = 0
+
+        # store the root in the nodes array
+        self.nodes.push_back(root)
+
+    cdef reset(self):
+        # clear the big memory users before we load an octree from memory
+        self.nodes.clear()
+        self._idx = np.zeros(1, dtype=np.int64)-1
+
+    cdef int build_tree(self, double * input_pos):
+        # generate an array to store the which particles are in each child oct
+        cdef np.int64_t * split_arr
+        split_arr = <np.int64_t*> malloc((self._num_children + 1) *
+                                                            sizeof(np.int64_t))
+
+        # loop through the octs in serial and process the nodes, i.e, sort the
+        # particles and create the children, which will increase the node.size
+        # if we make children, then we iterate through those children
+        cdef int num_nodes_processed = 0
+        while num_nodes_processed < self.nodes.size():
+            self.process_node(&self.nodes[num_nodes_processed], input_pos,
+                              split_arr)
+            num_nodes_processed += 1
```

Comment:
this might seem a little weird, but it prevents recursion which is good for keeping down the stack use. Recursively calling a function which use a bunch cdef's can get heavy. This processes a node, then comes back, checks if there are any unprocessed nodes, if so, we process another